python/ray/data/impl/arrow_block.py

```diff
@@ -489,6 +562,33 @@ def gen():
         return ret, ArrowBlockAccessor(ret).get_metadata(None, exec_stats=stats.build())
 
 
+def _is_column_extension_type(ca: "pyarrow.ChunkedArray") -> bool:
+    """Whether the provided Arrow Table column is an extension array, using an Arrow
+    extension type.
+    """
+    return isinstance(ca.type, pyarrow.ExtensionType)
+
+
+def _concatenate_extension_column(ca: "pyarrow.ChunkedArray") -> "pyarrow.Array":
+    """Concatenate chunks of an extension column into a contiguous array.
+
+    This concatenation is required for creating copies and for .take() to work on
+    extension arrays.
+    See https://issues.apache.org/jira/browse/ARROW-16503.
+    """
+    if not _is_column_extension_type(ca):
+        raise ValueError("Chunked array isn't an extension array: {ca}")
+
+    if ca.num_chunks == 0:
+        # No-op for no-chunk chunked arrays, since there's nothing to concatenate.
+        return ca
+
+    chunk = ca.chunk(0)
+    return type(chunk).from_storage(
+        chunk.type, pyarrow.concat_arrays([c.storage for c in ca.chunks])
```

Comment:
This will make a copy, it looks we do have copies when unpack arrow to ndarray.

Reply:
So yeah we currently always copy the slices (which [concatenates these chunks](https://github.com/ray-project/ray/blob/c045f3625edea65396487cd6ceff5e83363168e5/python/ray/data/impl/arrow_block.py#L603)) at [batch slicing time](https://github.com/ray-project/ray/blob/c045f3625edea65396487cd6ceff5e83363168e5/python/ray/data/dataset.py#L296) due to those zero-copy slice view serialization issues in Arrow.

I suppose that calling this zero-copy is disingenuous, I'll update that above example.

Reply:
I see, is that just the known Arrow issue? If so I think it has no problem to call it zero-copy (maybe just note a pending issue on Arrow side).

Reply:
Yep this is a known Arrow issue.